lib/rgb_matrix/animation/config.ex

```diff
@@ -97,42 +156,85 @@ defmodule RGBMatrix.Animation.Config do
       iex> RGBMatrix.Animation.Config.update_config(
       ...>   hue_wave_config,
       ...>   RGBMatrix.Animation.HueWave.Config.schema(),
-      ...>   %{"direction" => "left"}
-      ...> )
+      ...>   %{"direction" => "left"})
       %RGBMatrix.Animation.HueWave.Config{direction: :left, speed: 4, width: 20}
   """
-  @spec update_config(config :: t, schema :: any, params :: map) :: t
+  @spec update_config(
+          config :: t,
+          schema :: config_schema,
+          params :: update_params
+        ) :: t
   def update_config(config, schema, params) do
-    Enum.reduce(params, config, &update_field(&1, &2, schema))
+    params
+    |> Enum.reduce(config, fn param, config ->
+      cast_and_update_field(param, config, schema)
+    end)
   end
 
-  defp create_atom_key(key) when is_binary(key), do: String.to_existing_atom(key)
-  defp create_atom_key(key) when is_atom(key), do: key
+  @spec cast_and_update_field(
+          param :: {atom | String.t(), any},
+          config :: t,
+          schema :: config_schema
+        ) :: t
+  defp cast_and_update_field({key, value} = _param, config, schema) do
+    with key <- create_atom_key(key),
+         {:ok, %type_module{} = type} <- Keyword.fetch(schema, key),
+         {:ok, value} <- type_module.cast(type, value) do
+      Map.put(config, key, value)
+    else
+      error ->
+        field_warn(error, key, value)
```

Comment:
I do like the idea of providing as much information about the error as possible, so I'll work on this.

Reply:
:+1: 

Just to clarify, the main idea behind this is to simplify the API design; being able to provide additional information was a nice byproduct of restructuring the data. This change normalizes the shape of the return values and allows a developer to think about the return value in two steps:
1. Handling success/failure (`:ok | :error`)
2. Processing the relevant data

This may be more apparent when writing out explicit code that captures each type of return value. In the current codebase it may look something like this:

```ex
case cast() do
  {:ok, value} ->
  :error ->
  :cast_error ->
  :invalid_field ->
  :validation_error ->
end
```

Whereas in what I'm describing it may look more like this:

```ex
case cast() do
  {:ok, value} ->
  {:error, :invalid_field} ->
  {:error, :invalid_value} ->
  {:error, :value_too_small} ->
  {:error, :value_too_big} ->
end
```

Hopefully that's a better example of what I mean by making the API simpler. It's more consistent and predictable.

Reply:
Yeah, that makes parsing at a glance so much easier. I get what you're talking about.